File: day21/day21.cpp
#include <algorithm>
#include <fstream>
#include <regex>

#include <fmt/format.h>

using namespace std;

struct Instruction
{
	enum { SWP, SWL, ROT, ROP, REV, MOV, CNT } type;
	int a, b;

	void execute(string &str) const noexcept
	{
		switch(type)
		{
		case SWP:
			swap(str[a], str[b]);
			break;

		case SWL:
			for (auto& l: str)
			{
				if (l == a)
				{
					l = b;
				}
				else if (l == b)
				{
					l = a;
				}
			}
			break;

		case ROT:
			if (a)
			{
				// LEFT
				for (int i = 0; i < b; i++)
				{
					int v = str.front();
					str.erase(0, 1);
					str.push_back(v);
				}
			}
			else
			{
				// RIGHT
				for (int i = 0; i < b; i++)
				{
					int v = str.back();
					str.pop_back();
					str.insert(0, 1, v);
				}
			}
			break;

		case ROP: {
			size_t tmp = str.find(a);
			if (tmp != string::npos)
			{
				if (tmp >= 4)
				{
					tmp++;
				}
				tmp++;
				while (tmp-->0)
				{
					int v = str.back();
					str.pop_back();
					str.insert(0, 1, v);
				}
			}
			break;
		}

		case REV:
			for (size_t x = a, y = b; x < y; x++, y--)
			{
				swap(str[x], str[y]);
			}
			break;

		case MOV: {
			int tmp = str[a];
			str.erase(a, 1);
			str.insert(b, 1, tmp);
			break;
		}

		default:
			break;
		}
	}
};

istream& operator>>(istream& input, Instruction &i)
{
	static const regex rlst[] = {
		regex("swap position (\\d+) with position (\\d+)"),
		regex("swap letter ([a-z]) with letter ([a-z])"),
		regex("rotate (left|right) (\\d+) steps?"),
		regex("rotate based on position of letter ([a-z])"),
		regex("reverse positions (\\d+) through (\\d+)"),
		regex("move position (\\d+) to position (\\d+)"),
	};

	string line;
	if (!getline(input, line))
	{
		return input;
	}

	smatch sm;
	int j;
	for (j = 0; j < Instruction::CNT; j++)
	{
		if (regex_match(line, sm, rlst[j]))
		{
			break;
		}
	}
	i.type = static_cast<decltype(i.type)>(j);
	switch (j)
	{
	case Instruction::SWP:
	case Instruction::REV:
	case Instruction::MOV:
		i.a = stoi(sm[1]);
		i.b = stoi(sm[2]);
		break;

	case Instruction::SWL:
		i.a = sm[1].str()[0];
		i.b = sm[2].str()[0];
		break;

	case Instruction::ROT:
		i.a = sm[1] == "left";
		i.b = stoi(sm[2]);
		break;

	case Instruction::ROP:
		i.a = sm[1].str()[0];
		break;

	default:
		input.setstate(ios::failbit);
		return input;
	}
	return input;
}

static string scramble(string password, const vector<Instruction>& instructions)
{
	for (auto& i: instructions)
	{
		i.execute(password);
	}
	return password;
}

static string descramble(string password, const vector<Instruction>& instructions)
{
	string cur(password);
	sort(cur.begin(), cur.end());
	do
	{
		if (password == scramble(cur, instructions))
		{
			return cur;
		}
	} while (next_permutation(cur.begin(), cur.end()));

	return "";
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<Instruction> instructions;
	Instruction inst;
	while (input>>inst)
	{
		instructions.emplace_back(move(inst));
	}
	input.close();

	fmt::print("Part1: {}\n", scramble("abcdefgh", instructions));
	fmt::print("Part2: {}\n", descramble("fbgdceah", instructions));
	return 0;
}

File: day21/day21.py
#!/usr/bin/env python3

import re
import sys
from itertools import permutations

def swap_pos(lst, m):
    x, y = int(m[1]), int(m[2])
    lst[x], lst[y] = lst[y], lst[x]

def swap_letter(lst, m):
    x, y = m[1], m[2]
    for i in range(len(lst)):
        if lst[i] == x:
            lst[i] = y
        elif lst[i] == y:
            lst[i] = x

def rotate_dir(lst, m):
    d, steps = m[1], int(m[2])
    if d == "right":
        for i in range(steps):
            lst.insert(0, lst.pop())
    else:
        for i in range(steps):
            lst.append(lst.pop(0))

def rotate_pos(lst, m):
    letter = m[1]
    idx = lst.index(letter)
    if idx >= 4:
        idx += 1
    idx += 1
    rotate_dir(lst, ["", "right", idx])

def reverse(lst, m):
    x, y = int(m[1]), int(m[2])
    if x > y:
        x, y = y, x
    while x < y:
        lst[x], lst[y] = lst[y], lst[x]
        x += 1
        y -= 1

def move(lst, m):
    x, y = int(m[1]), int(m[2])
    lst.insert(y, lst.pop(x))

ops = [
    (re.compile(r"swap position (\d+) with position (\d+)"), swap_pos),
    (re.compile(r"swap letter ([a-z]) with letter ([a-z])"), swap_letter),
    (re.compile(r"rotate (left|right) (\d+) steps?"), rotate_dir),
    (re.compile(r"rotate based on position of letter ([a-z])"), rotate_pos),
    (re.compile(r"reverse positions (\d+) through (\d+)"), reverse),
    (re.compile(r"move position (\d+) to position (\d+)"), move),
]

def scramble(password, instructions):
    lst = list(password)
    for ins in instructions:
        for op, fun in ops:
            m = op.match(ins)
            if m:
                fun(lst, m)
                break
        else:
            print(ins, "not matched!")
            exit(1)

    return "".join(lst)

def descramble(password, instructions):
    for i in permutations(password):
        if scramble(i, instructions) == password:
            return "".join(i)

    return ""

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            instructions = f.read().strip().splitlines()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", scramble("abcdefgh", instructions))
    print("Part2:", descramble("fbgdceah", instructions))
